test/unit/client-daemonset.bats

```diff
@@ -109,6 +109,48 @@ load _helpers
   [ "${actual}" = "true" ]
 }
 
+@test "client/DaemonSet: grpc is enabled if connectInject.enabled=true" {
+  cd `chart_dir`
+  local actual=$(helm template \
+      -x templates/client-daemonset.yaml  \
+      --set 'connectInject.enabled=true' \
+      . | tee /dev/stderr |
+      yq '.spec.template.spec.containers[0].command | any(contains("grpc"))' | tee /dev/stderr)
+  [ "${actual}" = "true" ]
+}
+
+@test "client/DaemonSet: grpc is enabled if connectInject.enabled=-" {
+  cd `chart_dir`
+  local actual=$(helm template \
+      -x templates/client-daemonset.yaml  \
+      --set 'connectInject.enabled=-' \
+      . | tee /dev/stderr |
+      yq '.spec.template.spec.containers[0].command | any(contains("grpc"))' | tee /dev/stderr)
+  [ "${actual}" = "true" ]
+}
+
+@test "client/DaemonSet: grpc is enabled if connectInject.enabled=false but client.grpc=true" {
+  cd `chart_dir`
+  local actual=$(helm template \
+      -x templates/client-daemonset.yaml  \
+      --set 'connectInject.enabled=false' \
+      --set 'client.grpc=true' \
+      . | tee /dev/stderr |
+      yq '.spec.template.spec.containers[0].command | any(contains("grpc"))' | tee /dev/stderr)
+  [ "${actual}" = "true" ]
+}
+
+@test "client/DaemonSet: grpc is disabled if global.enabled=false" {
+  cd `chart_dir`
+  local actual=$(helm template \
+      -x templates/client-daemonset.yaml  \
+      --set 'global.enabled=false' \
+      --set 'client.enabled=true' \
+      . | tee /dev/stderr |
+      yq '.spec.template.spec.containers[0].command | any(contains("grpc"))' | tee /dev/stderr)
+  [ "${actual}" = "false" ]
+}
```

Comment:
Hmmm this test case is interesting. Currently the client daemon set template is disabled iff `global.enabled==false` AND `client.enabled=="-"`. So when client is explicitly enabled or disabled, the helm chart will ignore the global value. This means that the template will be rendered with whatever value `grpc` is set to in `values.yaml`, which happens to be `false` by default. Maybe a better test would be setting `client.enabled` to `-` and setting `client.grpc` to `true` and then checking that `grpc` is not present in the rendered template.

Reply:
> . Currently the client daemon set template is disabled iff global.enabled==false AND client.enabled=="-".

That seems like a bug. If client.enabled=false then it should be disabled.

> Maybe a better test would be setting client.enabled to - and setting client.grpc to true and then checking that grpc is not present in the rendered template.

In this case what's global.enabled set to? If true then we would expect client.grpc to be equal to true and if false then the whole template wouldn't render.

I'm not exactly sure what you're asking, maybe you can write the test you think this should be changed to?

Reply:
Sorry I misspoke, client template is disabled in two cases:
1. `client.enabled==false`
2. `global.enabled==false` AND `client.enabled=="-"`

I was mostly talking about the second case, but didn't mention it explicitly. 

I was a little confused by which logic this test is testing. In the test both `client.enabled` and `global.enabled` are set. The logic for the `global.enabled` value is never executed because it just short-circuits once it sees that `client.enabled` is true. 

There is already a test that tests that the template is rendered if `global.enabled` is `false` and `client.enabled` is `true`. I think we can assume that this will apply to all values set for the client and we don't have to test each value individually.

So maybe this test isn't necessary at all, unless you had some other intention behind it.